src/utils/cfgreader.py

```diff
@@ -0,0 +1,167 @@
+
+    def __init__(self, var_name: str, default_value, mask_value=False):
+        self.name = var_name
+        self.default_value = default_value
+        self.mask_value = mask_value
+
+    def read(self, rep: MaybeString):
+        """
+        Parse the value from its string representation as read from the
+        environment.
+
+        :param rep: the env value.
+        :return: the parsed value if the string rep holds a value,
+            the default value otherwise.
+        """
+        if self.has_value(rep):
+            return self._do_read(rep)
+        return self.default_value
+
+    def _do_read(self, rep: str):
+        pass
+
+
+class StrVar(EVar):
+    """
+    An env value parsed as a string. The string representation gets stripped
+    of any leading and trailing whitespace.
+    """
+
+    def _do_read(self, rep: str) -> str:
+        return rep.strip()
+
+
+class IntVar(EVar):
+    """
+    An env value parsed as an integer.
+    """
+
+    def _do_read(self, rep: str) -> int:
+        return int(rep)
+
+
+class BoolVar(EVar):
+    """
+    An env value parsed as a boolean. It evaluates to true just in case the
+    string representation, after trimming any leading and trailing whitespace,
+    is equal, ignoring case, to any of: 'true', 'yes', '1', 't'.
+    """
+
+    def _do_read(self, rep: str) -> bool:
+        return rep.strip().lower() in ('true', 'yes', '1', 't')
+
+
+class EnvReader:
+    """
+    Reads environment variables.
+    """
+
+    @staticmethod
+    def get_log_msg(var: EVar, value: MaybeString):
+        msgs = {
+            # (has value, mask value)
+            (True, True): "Env variable {name} set, using its value.",
+            (True, False): "Env variable {name} set to '{value}', " + \
+                           "using this value.",
+            (False, True): "Env variable {name} not set, " + \
+                           "using default value.",
+            (False, False): "Env variable {name} not set, " + \
+                            "using default value of: {default_value}"
+        }
+        return msgs[var.has_value(value), var.mask_value].format(
+            name=var.name, value=value, default_value=var.default_value
+        )
+
+    def __init__(self, var_store: dict = os.environ, log=None):
+        self.var_store = var_store
+        self.log = log if log else logging.getLogger(__name__).info
+
+    def read(self, var: EVar):
+        """
+        Read the specified environment variable and, if set, parse its
+        value; otherwise return the variable default.
+        Also log the value as read if it's not supposed to be masked.
+
+        :param var: the variable to read.
+        :return: the parsed value if the variable is set, the variable's
+            default otherwise.
+        """
+        env_value = self.var_store.get(var.name)
+        msg = self.get_log_msg(var, env_value)
+        self.log(msg)
+        return var.read(env_value)
+
+
+class YamlReader:
```

Comment:
I'd use a different name, what kind of yaml reader is this one?. 

Also, on a side note, I'd organise a bit better these ql-config-related files into a package with its specific purpose (config handling) to avoid creating a "utils" blackhole here.

Reply:
Yes, `YamlReader` isn't a great choice actually. I'll see if I can come up w/ something better. These classes are generic though, i.e. not tied in anyway to QL which is why I put them into util. Actually I've been (ab-)using the util namespace I reckon, I stashed away in there general purpose code that isn't tied to QL: `cfgreader`, `jsondict`, `kvt`, `streams`, and `timestr`. I'll find a better home for these guys when we refactor the timescale and crate backends.